homeassistant/components/sensor/__init__.py

```diff
@@ -179,13 +181,17 @@ def add_to_platform_start(
     ) -> None:
         """Start adding an entity to a platform.
 
+        Raise HomeAssistantError if sensor has the entity category config as a sensor cannot change something.
+
         Allows integrations to remove legacy custom unit conversion which is no longer
         needed without breaking existing sensors. Only works for sensors which are in
         the entity registry.
-
-        This can be removed once core integrations have dropped unneeded custom unit
-        conversion.
         """
+        if self.entity_category == EntityCategory.CONFIG:
+            raise HomeAssistantError(
+                f"Entity {type(self)} cannot be added as the entity category is set to config"
```

Comment:
Unfortunately, `entity_id` cannot used for the error message as it will be initialized later on.

Reply:
Moving it to `added_to_hass` to get the entity id